File: 03.DP动态规划/leetcode_1872_stoneGameVIII.py
from typing import List

class Solution:
    def stoneGameVIII(self, stones: List[int]) -> int:
        n = len(stones)
        tot = sum(stones)
        mx = tot
        for i in range(n-2, 0, -1):
            tot -= stones[i+1]
            if tot-mx > mx:
                mx = tot-mx
        return mx

        # ===================================

        # n = len(stones)
        # preSum = [0]*(n+1)
        # for i in range(n):
        #     preSum[i+1] = preSum[i]+stones[i]

        # mx = preSum[n]
        # for i in range(n-2, 0, -1):
        #     mx = max(mx, preSum[i+1]-mx)
        # return mx

        # ===================================
        
        # @cache
        # def dfs(pos):
        #     if pos >= n-1:
        #         return 0
        #     diff = -10**18
        #     for i in range(pos+1, n):
        #         diff = max(diff, preSum[i+1]-dfs(i))
        #     return diff
        
        # return dfs(0)

if __name__ == "__main__":
    s = Solution()
    stones = [-1,2,-3,4,-5]
    print(s.stoneGameVIII(stones))

File: 03.DP动态规划/leetcode_1872_stoneGameVIII.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int stoneGameVIII(vector<int>& stones) {
        int n = stones.size();
        int tot = 0;
        for (int i : stones) tot += i;
        int mx = tot;
        for (int i=n-2; i>=1; i--){
            tot -= stones[i+1];
            mx = max(tot-mx, mx);
        }
        return mx;
    }
};

int main(){
    Solution s;
    vector<int> stones = {-1,2,-3,4,-5};
    cout << s.stoneGameVIII(stones) << endl;
}
